### PR TITLE
[sgd] Add checkpointing

### PR DESCRIPTION
**Pull request by pschafhalter**
Adds checkpointing for distributed SGD.

`save_checkpoint`:
1. Get parameters from every worker
2. Save parameters to unique file for each model

`restore_checkpoint`:
1. Get parameters from files
2. Restore parameters for each model on all workers

First step for SGD fault tolerance #3016.

### CODE REVIEW

File: python/ray/experimental/sgd/sgd.py [hunk 1]
```diff
@@ -177,6 +180,33 @@ def warmup(self):
         ray.get([w.warmup.remote() for w in self.workers])
         logger.info("Warmup complete")
 
+    def save_checkpoint(self, path):
+        param_oids = [worker.get_params.remote() for worker in self.workers]
```
Comment (ericl) [inline python/ray/experimental/sgd/sgd.py]: Please use https://github.com/ray-project/ray/blob/master/python/ray/experimental/sgd/mnist_example.py#L115 instead.

You also don't need to fetch from each worker, just one will do.

File: python/ray/experimental/sgd/sgd.py [hunk 2]
```diff
@@ -177,6 +180,33 @@ def warmup(self):
...
+            params_list = ray.get(ready_oids)[0]
+            for params in params_list:
+                filename = os.path.join(path, "model_%d.pkl" % model_count)
+                with open(filename, "wb") as f:
+                    pickle.dump(params, f)
+                model_count += 1
+
+            ray.internal.free(ready_oids)
+
+    def restore_checkpoint(self, path):
```
Comment (ericl) [inline python/ray/experimental/sgd/sgd.py]: Use this: https://github.com/ray-project/ray/blob/master/python/ray/experimental/sgd/mnist_example.py#L116

File: python/ray/experimental/sgd/sgd_worker.py [hunk 3]
```diff
@@ -30,6 +30,7 @@ def __init__(self,
                  plasma_op=False):
         self.worker_index = worker_index
         assert num_devices > 0
+        self.num_devices = num_devices
```
Comment (ericl) [inline python/ray/experimental/sgd/sgd_worker.py]: This is not needed, all the devices share weights.

File: python/ray/experimental/sgd/sgd_worker.py [hunk 4]
```diff
@@ -263,3 +264,28 @@ def ip(self):
 
     def warmup(self):
         warmup()
+
+    def get_params(self):
```
Comment (ericl) [inline python/ray/experimental/sgd/sgd_worker.py]: https://github.com/ray-project/ray/blob/master/python/ray/experimental/sgd/mnist_example.py#L70 does the same thing

**Review by ericl — COMMENTED** [review]
I think you can do this much more simply by leveraging the built in API. For example (pseudocode):

in sgd.py:

def save_checkpoint(self):
    w0 = self.for_model(lambda m: m.get_variables().get_flat())
    with open(path) as f:
         f.write(w0)

def restore_checkpoint(self, path):
    w0 = open(path).read(
    sgd.foreach_model(lambda m: m.get_variables().set_flat(w0))

**Comment by richardliaw** [comment]
Yeah, built-in API would be good. For restoring, it's probably better to push weights into plasma first before putting it into the scope?

File: python/ray/experimental/sgd/sgd.py [hunk 5]
```diff
@@ -177,6 +180,33 @@ def warmup(self):
...
+        filenames = os.listdir(path)
```
Comment (stephanie-wang) [inline python/ray/experimental/sgd/sgd.py]: This doesn't seem to be used.

**Review by stephanie-wang — COMMENTED** [review]
I think this also needs some kind of heartbeat detection, but we can do that as a separate PR. You can expose a `ping()` method on the PS and worker actors that returns nothing. Then, on some interval, the driver should submit the method on all actors, and call `ray.wait` on them with some small timeout, maybe 0ms. If an object isn't ready, the actor is still alive, and you can try the same object ID on the next interval. If an object is ready, the actor is still alive, and you can submit another `ping()` task on the next interval. Else the object is ready, but you will get an exception because the actor has died.

This could be replaced with the signal API in #3624, but since we only need failure detection, this should be enough for now.

**Comment by pschafhalter** [comment]
@ericl @richardliaw thanks for the comments! I'll make the changes using the built-in API and should have those ready over the weekend (currently have low access to a computer).

@stephanie-wang I'll open a separate PR for the failure detection via `ping()` once this is merged. Thanks for looking this over!

**Comment by pschafhalter** [comment]
@ericl currently the `Model` abstract class doesn't include a `get_variables()` function -- it looks like this is only implemented in the MNIST model for now. What do you think the best way to solve this is? The following solutions might work:
1. Add `Model.get_variables()` which returns `ray.experimental.tfutils.TensorFlowVariables`. Unfortunately, this could make supporting other frameworks like PyTorch more difficult.
2. Add `Model.get_weights()` which returns a numpy array and `Model.set_weights()` which takes a numpy array. This would be more general, but requires implementing more methods when building a model.

**Comment by ericl** [comment]
(2) sounds fine, you can say these methods are optional and required for checkpointing

**Comment by pschafhalter** [comment]
This should be ready for review. The test failures look like they're unrelated.

I looked into using the tensorflow op to transfer weights between the model and plasma directly, but that might not be possible to do using the current get/set weights API which expects numpy arrays. Passing ObjectIDs wrapped in lists in get/set weights could solve this problem, but I'm not sure this is a good API change.

**Review by ericl — APPROVED** [review]
Look fine, some nits.

**Pull request merged**